libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/PlannedRouteRefreshController.kt

```diff
@@ -0,0 +1,148 @@
+package com.mapbox.navigation.core.routerefresh
+
+import androidx.annotation.VisibleForTesting
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.route.RouteRefreshOptions
+import com.mapbox.navigation.utils.internal.logI
+import kotlinx.coroutines.CoroutineScope
+
+@OptIn(ExperimentalPreviewMapboxNavigationAPI::class)
+internal class PlannedRouteRefreshController @VisibleForTesting constructor(
+    private val routeRefresherExecutor: RouteRefresherExecutor,
+    private val routeRefreshOptions: RouteRefreshOptions,
+    private val stateHolder: RouteRefreshStateHolder,
+    private val listener: RouteRefresherListener,
+    private val cancellableHandler: CancellableHandler,
+    private val retryStrategy: RetryRouteRefreshStrategy,
+) {
+
+    constructor(
+        routeRefresherExecutor: RouteRefresherExecutor,
+        routeRefreshOptions: RouteRefreshOptions,
+        stateHolder: RouteRefreshStateHolder,
+        scope: CoroutineScope,
+        listener: RouteRefresherListener,
+    ) : this(
+        routeRefresherExecutor,
+        routeRefreshOptions,
+        stateHolder,
+        listener,
+        CancellableHandler(scope),
+        RetryRouteRefreshStrategy(maxRetryCount = MAX_RETRY_COUNT)
+    )
+
+    private var paused = false
+    var routesToRefresh: List<NavigationRoute>? = null
+        private set
+
+    fun startRoutesRefreshing(routes: List<NavigationRoute>) {
+        cancellableHandler.cancelAll()
+        routesToRefresh = null
+        if (routes.isEmpty()) {
+            logI("Routes are empty", RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.reset()
+            return
+        }
+        val routesValidationResults = routes.map { RouteRefreshValidator.validateRoute(it) }
+        if (
+            routesValidationResults.any { it is RouteRefreshValidator.RouteValidationResult.Valid }
+        ) {
+            routesToRefresh = routes
+            scheduleNewUpdate(routes)
+        } else {
+            val message =
+                RouteRefreshValidator.joinValidationErrorMessages(
+                    routesValidationResults.mapIndexed { index, routeValidationResult ->
+                        routeValidationResult to routes[index]
+                    }
+                )
+            val logMessage = "No routes which could be refreshed. $message"
+            logI(logMessage, RouteRefreshLog.LOG_CATEGORY)
+            stateHolder.onFailure(logMessage)
```

Comment:
```suggestion
            stateHolder.onStarted()
            stateHolder.onFailure(logMessage)
```
wondering if this wouldn't be better to keep the state transition more predictable

Reply:
That's the previous behaviour. Didn't want to break it.
Discussing this in another thread.

Reply:
OK, that's not related.
It doesn't look worse and it's experimental. Why not.